swift/swift.go

```diff
@@ -201,56 +210,63 @@ func maybeNotFound(err error, format string, arg ...interface{}) error {
 	return errors.Newf(err, format, arg...)
 }
 
-type oreadseeker struct {
+type readSeeker struct {
 	c             *Client
 	containerName string
 	objectName    string
 	length        int64
 	pos           int64
+	etag          string
 }
 
-// Read reads up to len(b) bytes from the oreadseeker. It returns the number of
+var acceptableReadStatuses = []int{http.StatusOK, http.StatusPartialContent, http.StatusPreconditionFailed}
+
+// Read reads up to len(b) bytes from the readSeeker. It returns the number of
 // bytes read and any error encountered. At end of stream, Read returns no EOF
 // on the first EOF and EOF after that.
-func (o *oreadseeker) Read(p []byte) (n int, err error) {
-	if o.pos == o.length {
+func (o *readSeeker) Read(p []byte) (n int, err error) {
+	if o.pos >= o.length {
 		return 0, io.EOF
 	}
-	end := int64(len(p)) + o.pos
+	end := o.pos + int64(len(p))
+	if end > o.length {
+		p = p[:o.length-o.pos]
+		end = o.pos + int64(len(p))
+	}
 	header := make(http.Header)
 	header.Add("Range", fmt.Sprintf("bytes=%d-%d", o.pos, end))
+	if o.etag != "" {
+		// Ensure that the read fails if the file has
+		// been written to since we opened it.
+		header.Set("If-Match", o.etag)
+	}
 	requestData := goosehttp.RequestData{
+		// Signal to the client that we want the reader back
+		// by placing a dummy reader there.
+		// (RespReader gets replaced).
 		RespReader:     &emptyReadCloser,
 		ReqHeaders:     header,
-		ExpectedStatus: []int{http.StatusOK, http.StatusPartialContent},
+		ExpectedStatus: acceptableReadStatuses,
```

Comment:
I pulled it into a package variable to avoid allocating a new slice every time. 
I don't see too much difference in the names, so *shrug* done.